matrix_content_scanner/crypto.py

```diff
@@ -0,0 +1,91 @@
+#  Copyright 2022 The Matrix.org Foundation C.I.C.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import json
+import logging
+from typing import TYPE_CHECKING
+
+from olm.pk import PkDecryption, PkDecryptionError, PkMessage
+
+from matrix_content_scanner.utils.constants import ErrCode
+from matrix_content_scanner.utils.errors import ContentScannerRestError
+from matrix_content_scanner.utils.types import JsonDict
+
+if TYPE_CHECKING:
+    from matrix_content_scanner.mcs import MatrixContentScanner
+
+
+logger = logging.getLogger(__name__)
+
+
+class CryptoHandler:
+    """Handler for handling Olm-encrypted request bodies."""
+
+    def __init__(self, mcs: "MatrixContentScanner") -> None:
+        key = mcs.config.crypto.pickle_key
+        path = mcs.config.crypto.pickle_path
+        try:
+            # Try reading the pickle from disk.
+            with open(path, "r") as fp:
+                pickle = fp.read()
+
+            # Create a PkDecryption object with the content and key.
+            self._decryptor: PkDecryption = PkDecryption.from_pickle(
+                pickle=pickle.encode("ascii"),
+                passphrase=key,
+            )
+
+            logger.info("Loaded Olm pickle from %s", path)
+        except FileNotFoundError:
+            # If the pickle file doesn't exist, try creating it.
+            logger.info(
+                "Olm pickle not found, generating one and saving it at %s", path
+            )
+
+            self._decryptor = PkDecryption()
+            pickle_bytes = self._decryptor.pickle(passphrase=key)
+
+            with open(path, "w+") as fp:
+                fp.write(pickle_bytes.decode("ascii"))
+
+        self.public_key = self._decryptor.public_key
```

Comment:
ok, so something that confused me on my first review: the logs and comments here talk about loading the pickle, but don't explain what that pickle is actually doing.

So, rather than taking about pickles, can we instead talk about private keys. "First try to load our keypair from the pickle file" ... "If the pickle file doesn't exist, create a new keypair and save it in a new pickle". (and update the log statement accordingly)

Also: where is the config file documented? It should say something about the fact a new key is magically created if the pickle doesn't exist.

(TBH I slightly wonder if silently creating a new key is a footgun. I might prefer to *know* that my keyfile has gone missing, rather than suddenly getting new keys?)

Reply:
> Also: where is the config file documented? It should say something about the fact a new key is magically created if the pickle doesn't exist.

Hah, I forgot to port that bit :facepalm: thanks for noticing!

> (TBH I slightly wonder if silently creating a new key is a footgun. I might prefer to know that my keyfile has gone missing, rather than suddenly getting new keys?)

fwiw the point of this is to mirror what the current content scanner does, but I guess the creation of the file can be moved to a script.

Reply:
> Also: where is the config file documented? It should say something about the fact a new key is magically created if the pickle doesn't exist.

Oh wait no, that bit does exist: https://github.com/matrix-org/matrix-content-scanner-python/blob/b7563c3e6e5a6fe98ea9578623b51e3a7a77f91f/config.sample.yaml#L101-L109

(but yeah it should mention if it's creating the file)